### ISSUE TITLE
[feature] compress then/or encrypt

### ISSUE THREAD

**Issue by invictus-0x90** [TheWover/donut#12]
It would be nice to have compression as a feature. This could either be implemented as compress or encrypt, or compress and encrypt. Compressing an already encrypted payload is not viable.

I think compressapi would probably work: https://docs.microsoft.com/en-us/windows/win32/api/compressapi/nf-compressapi-compress

I'm happy to try to do this myself, but could do with some guidance on where to implement. Initial thoughts are that this wouldn't neccessarily be easy, would probably require some stub that isn't compressed who's sole purpose is to decompress the actual payload and inject it into memory.

As a workaround, currently we would have to generate shellcode from donut (with encryption disabled). Compress the shellcode file. Put this data into another executable that decompresses and injects into memory, and then finally use donut on this new file to get the final shellcode.

**Comment by odzhan** [comment]
Hi. Early in dev, compressing the payload before encryption using the built-in API for windows was considered, although after searching for Linux implementations, it got put on the back burner. I would prefer to compress first, then encrypt. Do you know of open source implementations in C that are compatible with Windows API? I haven't looked at the internals yet to see what potential problems exist.

**Comment by odzhan** [comment]
After looking at the internals, the compression algorithms used by RtlCompressBuffer can probably be implemented on Linux without any problem. Just to clarify. Are you suggesting the entire shellcode, including module be compressed or just the module?

**Comment by invictus-0x90** [comment]
I'm not too familiar with the difference between the entire shellcode and "just the module", I guess whatever gives the best results or is easiest. I'm mainly requesting this so we can use it with https://github.com/FSecureLABS/C3. The executables and dlls from that project can be quite large >1mb. So compression would really help in getting more manageable payloads

**Comment by invictus-0x90** [comment]
Just to make sure we're on the same page, the full output wouldn't be compressed right? The idea would be that the generated binary from donut decompresses the buffer in memory then executes as per usual.

**Comment by invictus-0x90** [comment]
Yeh that makes perfect sense thanks!

**Comment by TheWover** [comment]
I'll just reverse engineer `RtlCompressBuffer` in Ghidra to get the compression algorithm right. Easiest way to make sure the algorithm is 1-1.

**Comment by invictus-0x90** [comment]
No complaints on my side!

**Comment by TheWover** [comment]
So. we looked into this. Turns out, the LZ algorithm used by `RtlCompressBuffer` is proprietary. The Xpress algorithm is complicated and using existing libraries would require a significant increase in the size and complexity of Donut. Additionally, the LZ algorithm is proprietary. We could re-implement it in C, but would rather not deal with that potential mess. Instead, we are going to implement a minimal compression directly in Donut. We found a minimal implementation of the UPX compression algorithm that should take up <= 50 lines of code. Unless there are any objections, we will go with that. It would have the added benefit that API Hooking would not notice any calls to Windows compression APIs. But it may add some additional risk if memory scanners look for UPX-packed data in memory.
